tallypush/tally.go

```diff
@@ -93,14 +95,47 @@ type histogram struct {
 
 	// lasts keep the last value pushed to tally per histogram bucket.  This
 	// defaults to zero.
-	lasts map[int64]int64
+	lasts       []int64
+	bucketValue []int64
 }
 
+// Set is log(n) because it performs binary search to find the index that bucket belongs to
 func (th *histogram) Set(bucket int64, total int64) {
-	delta := total - th.lasts[bucket]
-	th.lasts[bucket] = total
+	index := sort.Search(len(th.lasts), func(i int) bool {
+		return th.bucketValue[i] >= bucket
+	})
+
+	th.ensureBucket(index, bucket, false)
+	th.recordValue(index, bucket, total)
+}
+
+func (th *histogram) ensureBucket(index int, bucket int64, panicOnError bool) {
+	// this shouldn't happen, but we should support it, to be fallback compatible
```

Comment:
is this comment for the whole method or the first case? If it is for the method, move to before the method. If it is for case1, move after the `switch`